text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,607 @@
+
+Invocations of overloaded operators with at least one non-primitive argument were considered as valid targets, but were
+rejected on grounds of being too error-prone. In any case, these can still be called as methods.
+
+## Type checking
+[typechecking]: #typechecking
+A `become` statement is type-checked like a `return` statement, with the added restriction of exactly matching the
+function signatures between caller and callee. Additionally, the caller and callee **must** use the same calling
+convention.
+
+## Borrowchecking and Runtime Semantics
+[semantics]: #semantics
+A `become` expression acts as if the following events occurred in-order:
+
+1. All variables that are being passed by-value are moved to temporary storage.
+2. All local variables in the caller are destroyed according to usual Rust semantics. Destructors are called where
+   necessary. Note that values moved from in step 1 are _not_ dropped.
+3. The caller's stack frame is removed from the stack.
+4. Control is transferred to the callee's entry point.
+
+This implies that it is invalid for any references into the caller's stack frame to outlive the call. The borrow checker ensures that none of the above steps will result in the use of a value that has gone out of scope.
+
+As `become` is always in a tail position (due to being used in place of `return`), this requirement for TCO is already
+fulfilled.
+
+Example:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    become y(a)
+}
+```
+
+Will be desugared in the following way:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    let _tmp = a;
+    drop(b);
+    become y(_tmp)
+}
+```
+
+## Implementation
+[implementation]: #implementation
+
+A now six years old implementation for the earlier mentioned
+[RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md) can be found at
+[DemiMarie/rust/tree/explicit-tailcalls](https://github.com/DemiMarie/rust/tree/explicit-tailcalls).
+A new implementation is planned as part of this RFC.
+
+The parser parses `become` exactly how it parses the `return` keyword. The difference in semantics is handled later.
+
+During type checking, the following are checked:
+
+1. The target of the tail call is, in fact, a simple call.
+2. The target of the tail call has the proper ABI.
+
+Later phases in the compiler assert that these requirements are met.
+
+New nodes are added in HIR and THIR to correspond to `become`. In MIR, the function call is checked that:
+1. The returned value is directly returned.
+2. There are no cleanups.
+3. The basic block being branched into has length zero.
+4. The basic block being branched into terminates with a return.
+
+If these conditions are fulfilled the function call following `become` is flagged to indicate the TCO requirement. This
+flag is then propagated to the corresponding backend. In the backend, there is an additional check if TCO can be
+performed.
```

Comment:
It might make more sense to introduce a `TailCall` MIR node which combines a call and a return. This guarantees that nothing is ever inserted between the call and become.

Reply:
Added this suggestion here: https://github.com/rust-lang/rfcs/pull/3407/commits/53a15cc486d512cde20e93e7bb435b7b45e06c8f